Add tombstones in mappings

AmitPhulera opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->

As part of ongoing elastic search upgrade work, this PR adds `__is_tombstone__` in every index mapping.

## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
https://dimagi-dev.atlassian.net/browse/SAAS-14007

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->


### Migrations
<!-- Delete this section if the PR does not contain any migrations -->
<!-- https://commcare-hq.readthedocs.io/migrations_in_practice.html -->
- [x] The migrations in this code can be safely applied first independently of the code

<!-- Please link to any past code changes that are coordinated with this migration -->

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [ ] This PR can be reverted after deploy with no further considerations
The migrations in the PR should be reverted before reverting the PR.

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

AmitPhulera committed:
add __is_tombstone__ in all mappings

AmitPhulera committed:
isort

AmitPhulera committed:
add migration to update mappings

AmitPhulera committed:
check for tombstone value before raising

AmitPhulera commented on corehq/apps/es/migrations/0001_add_tombstone_in_mappings.py:
```diff
@@ -0,0 +1,30 @@
+    ]
+
+    operations = [
+        update_es_mapping(CASE_SEARCH_ALIAS),
```
I think this will apply the entire mapping changes and I believe we would not want that? Is there is anything built in HQ for this? I can also use the normal es API to update the specific part.
Looking for your suggestions @calellowitz @joxl

joxl replied:
>

Correct. The ticket I'm currently working ([SAAS-14003](https://dimagi-dev.atlassian.net/browse/SAAS-14003)) on will add this.

AmitPhulera committed:
update migration name in migration.lock

calellowitz commented on migrations.lock:
```diff
@@ -343,6 +343,8 @@ enterprise
  0003_enterprisepermissions_modify_account
  0004_enterprisemobileworkersettings
+es
+ 0001_initial
```
should this name reflect the actual migration name?

AmitPhulera replied:
Yeah. It should be, have updated it in https://github.com/dimagi/commcare-hq/pull/32393/commits/02977040008d2aa5032d6cb5028c8134a3ef2a4a

calellowitz reviewed (approved):
Not blocking but curious about the lock file

AmitPhulera committed:
Merge remote-tracking branch 'origin/master' into ap/add-tombstones-in-mappings

AmitPhulera committed:
Merge branch 'ap/add-tombstones-in-mappings' of https://github.com/dimagi/commcare-hq into ap/add-tombstones-in-mappings

AmitPhulera committed:
remove previous migration

AmitPhulera committed:
run make_elastic_migration

./manage.py make_elastic_migration --name add_tombstones -u apps:__is_tombstone__ -u case_search:__is_tombstone__ -u cases:__is_tombstone__ -u domains:__is_tombstone__ -u forms:__is_tombstone__ -u groups:__is_tombstone__ -u sms:__is_tombstone__ -u users:__is_tombstone__

AmitPhulera committed:
update migration.lock

joxl commented on corehq/apps/es/migrations/0001_add_tombstones.py:
```diff
@@ -0,0 +1,69 @@
+            },
+        ),
+        corehq.apps.es.migration_operations.UpdateIndexMapping(
+            name='case_search_2018-05-29',
```
This will need to conditionally come from Django settings. See [here](https://github.com/dimagi/commcare-hq/blob/c201e2f5c20/corehq/apps/es/case_search.py#L158).

joxl commented on corehq/apps/es/migrations/0001_add_tombstones.py:
```diff
@@ -0,0 +1,69 @@
+            },
+        ),
+        corehq.apps.es.migration_operations.UpdateIndexMapping(
+            name='xforms_2016-07-07',
```
Same as above. See [forms.py](https://github.com/dimagi/commcare-hq/blob/c201e2f5c20/corehq/apps/es/forms.py#L67).

AmitPhulera committed:
get index name from settings where required

joxl commented:
This PR should be merged into `ap/bootstrap-es-indexes` (#32504) instead of `master` because it depends on that migration.

AmitPhulera commented:
Yep. I am holding the merge. I would need to update the migration name when the index bootstrap PR will get merged.

AmitPhulera committed:
merge commit

AmitPhulera committed:
regenerated migrations after merge

joxl reviewed (approved):
Only an inconsequential nit. Looks good.

joxl commented:
If you haven't already give this some thought, I would recommend the following steps when you're ready to merge as a final "pre-merge sanity check":

1. Deploy to staging.
2. Observe the deploy output to verify the migration happens as expected (the `migrate` output will contain diffs).
3. If that result is as expected, merge.

AmitPhulera committed:
feedback from commit

Co-authored-by: Joel M <jmiller@dimagi.com>

AmitPhulera commented:
Tested on staging and the diff looked correct. Going to merge this now.
